### DISCUSSION TITLE
Compile JS into a wasm module that can be called from a custom hostruntime

### DISCUSSION THREAD

**Discussion by ullgren** [WasmEdge/WasmEdge#2085] [Q&A] [answered]
Hello,

I'm quite new to wasm so this might be a completely noob question, in that case sorry about that :-)

I'm looking for a way to recreate this example, https://wasmedge.org/book/en/sdk/go/function.html but with a compiled wasm file written in javascript instead of rust. That is I would like to use the same go host application, just replace the WASM file with one build from JS source.
Is this possible with the wasmedge tool chain ? If not any suggestions on resources would be appreciated. 

I know about running embedded JS using the wasmedge, however what I want to do is have a host runtime that loads a WASM file and calls exported functions.

Best regards
Pontus Ullgren

**Comment by alabulei1** [comment, marked as answer]
Hi @ullgren 

The answer is Yes! You could embed WasmEdge into your go host application and then execute js applications in WasmEdge. See a code example here. https://github.com/Edgenesis/wasm-shifu-demo

**Reply by ullgren** [reply]
Thanks, I will look into this approach.

**Reply by ullgren** [reply]
Hello again,

The wasm-shifu-demo does no really do what I'm looking for. 
Lets say that I have a javascript file with the following functions defined.

```
function greet(name){
    return "Hello " + name;
}

function reverse(name) {
    return name.split("").reverse().join("");
}
```

Is there anyway from the `JsValue` returned from  `ctx.eval_global_str` to get one of the functions and execute it.

**Reply by juntao** [reply]
Hello @ullgren 

WasmEdge runs JS in a QuickJS runtime, which is itself compiled to Wasm. It does not "compile JS to Wasm". 

So, in order for the JS program to call host functions (like our node server, fetch(), and Tensorflow JS API), you will need to first extend and recompile QuickJS to support those host functions and then expose those host functions as JS APIs. 

For example, the `hi()` function below is implemented in Rust and exposed as a JS API. You could call a host function from `hi()` using Rust. 

https://wasmedge.org/book/en/write_wasm/js/rust.html#create-a-javascript-function-api

**Reply by ullgren** [reply]
Thanks for your quick reply.
